include/onnxruntime/core/framework/op_kernel.h

```diff
@@ -63,13 +72,46 @@ class OpKernel {
   //                   The kernel is responsible for keeping the packed data and related metadata if is_packed is true,
   //                   and the original initialized constant tensor will be released and not accessible anymore in
   //                   the Compute function.
-  virtual Status PrePack(const Tensor& /*tensor*/, int /*input_idx*/, bool& is_packed) {
+  // @param prepacked_weight_for_caching: A packed weight instance will be provided to the kernel IF the pre-packed weights
+  //                                      are meant to be packed. This needs to be filled in by the kernel and a boolean flag
+  //                                      (is_filled_) needs to be set to true.
+  // @param alloc: The kernel's PrePack() method MUST use this allocator for allocating the pre-packed
+  //               weights' buffers. The alloc that the PrePack() method will receive will be either
+  //               the allocator tied to the session if the kernel owns the pre-packed buffer or an
+  //               allocator shared between sessions if the pre-packed buffer is to be shared across sessions
+  //               (i.e.) the kernel does not own the buffer.
+
+  virtual Status PrePack(const Tensor& /*tensor*/, int /*input_idx*/, /*out*/ bool& is_packed,
+                         /*out*/ PrepackedWeight* /*prepacked_weight_for_caching*/,
+                         AllocatorPtr /*alloc*/) {
     is_packed = false;
     return Status::OK();
   }
 
+  // Override this function to use provided pre-packed weight.
+  // Status StorePrePackedWeight(const PrepackedWeight& prepacked_weight,
+  //                                        int input_idx,
+  //                                        /*out*/ bool& stored_weight) {
+  //     stored_weight = true;
+  //     // LOGIC TO USE PROVIDED PREPACKED WEIGHT AND ASSIGN IT TO this.buffer_
+  //     return Status::OK();
+  //   }
+  // Please refer to MatMulIntegerToFloatBase for a complete example
+  // @param prepacked_weight: The pre-packed weight to be used by this kernel
+  // @param input_idx: The input index of the tensor in this kernel
+  // @param stored_weight: Boolean flag set by the kernel implementation indicating
+  // that the provided weight has been stored by the kernel.
+  virtual Status StorePrePackedWeight(const PrepackedWeight& /*prepacked_weight*/,
+                                      int /*input_idx*/,
```

Comment:
Feels like this separation of PrePack and StorePrePackedWeight is a little confusing. 

Would it make sense to have a wrapper class around a PrepackedWeight that is passed in the existing PrePack to simplify? StorePrePackedWeight would go away.

Something along the lines of kernel does the initial pre packing, passes that in to the wrapper class which handles the hash lookup, cache usage etc. that is currently in SessionState::PrepackConstantInitializedTensors, and returns a previously cached value if the kernel should use that.

Reply:
The idea behind the separation is to keep the kernels unaware of the existence of a cache. A layer above the kernel (session state) would fetch the pre-packed weights from the kernel, see if it exists in the cache based on the hash of the pre-packed weights memory (if cache is enabled) and then store either the cached value or the fetched value in the kernel. If the cache is disabled, it would just store the fetched value. From the kernel's pov, it doesn't care what is getting stored.

Reply:
Either way they're somewhat aware. PrePack now takes prepacked_weight_for_caching and needs to use that correctly. 

If you have a helper that plugs in the packing at that point you don't need StorePrePackedWeight so there's only one thing for the kernel to implement. Whether it's writing to prepacked_weights_for_caching or calling a helper is pretty similar isn't it?

Reply:
prepacked_weight_for_caching is a bad name. From the kernel's pov, they're just filling in a container of pre-packed weights without the knowledge of whether it's cached or not.

Reply:
Discussed with Pranav offline. Suggestion is to simplify the 'store' to just take a std::vector<void*> and the input_idx.

Kernel should have something equivalent to this:
```
  void* packed_weight_;
  BufferUniquePtr owned_packed_weight_;
```

In PrePack it should allocate and set packed_weight_
```
  owned_packed_weight_ = BufferUniquePtr(data, BufferDeleter(alloc));
  packed_weight_ = owned_packed_weight_;
```

packed_weight_ can be replaced by what is currently StorePrePackedWeight. 

```
Status StorePrePackedWeights(const std::vector<void*>& shared_packed_weights, int input_idx) { 
  ...
  if (shared_packed_weights[0] != nullptr) {
    packed_weight_ = shared_packed_weights[0];
    owned_packed_weight_ = nullptr; // release
  }
  ...
}
```

Not sure there's value in setting 'stored_weight' either. Why does the session care if the kernel did or did not store what it gave it? There's no ownership transfer so it seems like unnecessary code.

Reply:
I can rename this prepacked_weight similar to the naming in StorePrePackedWeight()

Reply:
As synced offline, can we just have one kernel member instead of 2 per pre-packed weight like this:

  void* packed_weight_;
  BufferUniquePtr owned_packed_weight_; 

? If a kernel has multiple pre-packed weights, this quickly clutters the kernel. 

We can just has the BufferUniquePtr and in StorePrepackedWeights, we can pass in a vector of BufferUniquePtrs that just get moved into the kernel member.